### PR TITLE
Analyzer: Add Feature - Add basic analysis features

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Add basic analysis features.

**Major Revision**
- Add statistics, correlations of the raw data
- Add numeric outlier detection(inter_quartile_range)
- Add boxplot for selected metric

### CODE REVIEW

File: superbench/analyzer/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""Exposes interfaces of SuperBench Analyzer."""
+
+__all__ = ['data_analysis']
```
Comment (guoshzhao) [inline superbench/analyzer/__init__.py:6]: Looks there have no difference whether put `data_analysis` into `__all__` or not. If you import it in this way:
`import superbench.analyzer.data_analysis as data_analysis `

File: superbench/analyzer/data_analysis.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+import pandas as pd
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+from superbench.common.utils import logger
+
+
+def statistic(raw_data_df):
+    """Get the statistics of the raw data, including count, mean, std, min, max, 25%, 50%, 75%.
```
Comment (guoshzhao) [inline superbench/analyzer/data_analysis.py:14]: I think 95th value is an important value, especially for latency analysis, could we add it?

File: superbench/analyzer/data_analysis.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    Args:
+        raw_data_df (DataFrame): raw data
+
+    Returns:
+        DataFrame: data statistics
+    """
+    if len(raw_data_df) == 0:
+        return pd.DataFrame()
+    data_statistics_df = raw_data_df.describe()
```
Comment (guoshzhao) [inline superbench/analyzer/data_analysis.py:24]: Could we check the input type?

File: superbench/analyzer/data_analysis.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    statistics_error = []
+    for column in list(raw_data_df.columns):
+        if column not in list(data_statistics_df.columns) and not raw_data_df[column].isnull().all():
```
Comment (guoshzhao) [inline superbench/analyzer/data_analysis.py:27]: Is the logical expression wrong? I think you want to confirm whether all values in the missing column is null? It should be
`and raw_data_df[column].isnull().all()`  not `and not raw_data_df[column].isnull().all()`
BTW, looks the UT can't test this code path.

File: superbench/analyzer/data_analysis.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            statistics_error.append(column)
+    if statistics_error:
+        logger.warning(
+            'DataAnalyzer: [{}] is missing in statistics results.'.format(','.join(str(x) for x in statistics_error))
+        )
+    return data_statistics_df
+
+
+def inter_quartile_range(raw_data_df):
+    """Get the IQR outlier detection bound, including mild and extreme outlier upper and lower value and bound.
```
Comment (guoshzhao) [inline superbench/analyzer/data_analysis.py:37]: The docstring of this function is ambiguous...

File: superbench/analyzer/__init__.py (hunk 1 above)
Comment (yukirora) [reply superbench/analyzer/__init__.py]: removed this file

File: superbench/analyzer/data_analysis.py (hunk 2 above)
Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: sure, added

File: superbench/analyzer/data_analysis.py (hunk 3 above)
Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: sure

File: superbench/analyzer/data_analysis.py (hunk 4 above)
Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: it's not logical expression wrong, I mean for the columns that may not be float type, and UT has tested for this

File: superbench/analyzer/data_analysis.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        if column not in list(data_corr_df.columns) and not raw_data_df[column].isnull().all():
+            statistics_error.append(column)
+    if statistics_error:
+        logger.warning(
+            'DataAnalyzer: [{}] is missing in correlation results.'.format(','.join(str(x) for x in statistics_error))
+        )
+    return data_corr_df
+
+
+def creat_boxplot(raw_data_df, columns):
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: I add a line to save the fig to path, now it works well

File: superbench/analyzer/data_analysis.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+import numpy as np
+import pandas as pd
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+from superbench.common.utils import logger
+
+
+def statistic(raw_data_df):
+    """Get the statistics of the raw data, including count, mean, std, min, max, 25%, 50%, 75%, 95%.
```
Comment (cp5555) [inline superbench/analyzer/data_analysis.py:15]: Would you please add 1%, 5%, 99%?

File: superbench/analyzer/data_analysis.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+    Returns:
+        DataFrame: data statistics
+    """
+    data_statistics_df = pd.DataFrame()
+    if not isinstance(raw_data_df, pd.DataFrame):
+        logger.error('DataAnalyzer: the type of raw data is not pd.DataFrame')
+        return data_statistics_df
+    if len(raw_data_df) == 0:
+        return data_statistics_df
```
Comment (cp5555) [inline superbench/analyzer/data_analysis.py:28]: Do we need the table loc?

File: superbench/analyzer/data_analysis.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    if not isinstance(columns, list):
+        logger.error('DataAnalyzer: the type of columns should be list.')
+        return
+    try:
+        data_columns = raw_data_df.columns
+        for column in columns:
+            if column not in data_columns or raw_data_df[column].dtype is not np.dtype('float'):
+                logger.warning('DataAnalyzer: invalid column {} for boxplot.'.format(column))
+                columns.remove(column)
+        temp_raw_data_df = raw_data_df.fillna(value=raw_data_df.mean())
```
Comment (cp5555) [inline superbench/analyzer/data_analysis.py:155]: it's incorrect. Please remove it.

File: superbench/analyzer/data_analysis.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            plt.subplot(n, 1, i + 1)
+            sns.boxplot(x=columns[i], data=temp_raw_data_df, orient='h')
+        plt.subplots_adjust(hspace=1)
+        plt.savefig(output_dir + '/boxplot.png')
+        plt.show()
+    except Exception as e:
+        logger.error('DataAnalyzer: creat_boxplot failed, msg: {}'.format(str(e)))
+
+
+def export_baseline(raw_data_df, output_dir):
```
Comment (cp5555) [inline superbench/analyzer/data_analysis.py:168]: generate_baseline

Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: done

File: superbench/analyzer/data_analysis.py (hunk 9 above)
Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: done

File: superbench/analyzer/data_analysis.py (hunk 7 above)
Comment (yukirora) [reply superbench/analyzer/data_analysis.py]: added

File: superbench/analyzer/data_analysis.py (hunk 8 above)
Comment (yukirora) [reply superbench/analyzer/data_analysis.py:31]: I think maybe it's ok to keep empty all

Comment (cp5555) [reply superbench/analyzer/data_analysis.py:31]: Thanks

**Review by cp5555 — APPROVED** [review]

File: .azure-pipelines/cuda-unit-test.yml [hunk 11]
```diff
@@ -30,7 +30,7 @@ steps:
   - script: |
       SB_MICRO_PATH=$PWD python3 setup.py test
     displayName: Run unit tests
-    timeoutInMinutes: 10
+    timeoutInMinutes: 15
```
Comment (cp5555) [inline .azure-pipelines/cuda-unit-test.yml:33]: Why need to change it?

Comment (yukirora) [reply .azure-pipelines/cuda-unit-test.yml:33]: the timeout is not enough because I add some python package installation

**Review by guoshzhao — APPROVED** [review]

File: superbench/analyzer/data_analysis.py (hunk 4 above)
Comment (guoshzhao) [reply superbench/analyzer/data_analysis.py]: Got it.

**Pull request merged**